gettsim/dag.py

```diff
@@ -1,116 +1,243 @@
-        of its data dependencies.
+        The complete DAG of the tax and transfer system.
 
     """
-    dag = nx.DiGraph()
+    # Construct DAG from dictionary.
+    functions_arguments_dict = {
+        name: get_names_of_arguments_without_defaults(function)
+        for name, function in functions.items()
+    }
+    dag = nx.DiGraph(functions_arguments_dict).reverse()
+
+    # Save functions in DAG.
     for name, function in functions.items():
-        dag.add_node(name, function=function)
-        for dependency in inspect.getfullargspec(function).args:
-            attr = (
-                {"function": functions.get(dependency)}
-                if dependency in functions
-                else {}
-            )
-            dag.add_node(dependency, **attr)
-            dag.add_edge(dependency, name)
+        dag.nodes[name]["function"] = function
 
     return dag
 
 
-def prune_dag(dag, targets):
-    """Prune the dag.
+def _fail_if_targets_not_in_functions(functions, targets):
+    """Fail if targets are not in functions.
+
+    Parameters
+    ----------
+    functions : dict of callable
+        Dictionary containing functions to build the DAG.
+    targets : list of str
+        The targets which should be computed. They limit the DAG in the way that only
+        ancestors of these nodes need to be considered.
+
+    Raises
+    ------
+    ValueError
+        Raised if ``targets`` are not in functions.
+
+    """
+    targets_not_in_functions = set(targets) - set(functions)
+    if targets_not_in_functions:
+        formatted = format_list_linewise(targets_not_in_functions)
+        raise ValueError(
+            f"The following targets have no corresponding function:\n{formatted}"
+        )
+
+
+def _limit_dag_to_targets_and_their_ancestors(dag, targets):
+    """Limit DAG to targets and their ancestors.
 
     Parameters
     ----------
     dag : networkx.DiGraph
-        The unpruned DAG.
-    targets : list
-        Variables of interest.
+        The complete DAG.
+    targets : list of str
+        Variables which should be computed.
 
     Returns
     -------
     dag : networkx.DiGraph
         Pruned DAG.
 
     """
-    all_ancestors = set(targets)
+    used_nodes = set(targets)
     for target in targets:
-        all_ancestors = all_ancestors | set(nx.ancestors(dag, target))
+        used_nodes = used_nodes | set(nx.ancestors(dag, target))
 
     all_nodes = set(dag.nodes)
 
-    redundant_nodes = all_nodes - all_ancestors
+    unused_nodes = all_nodes - used_nodes
 
-    dag.remove_nodes_from(redundant_nodes)
+    dag.remove_nodes_from(unused_nodes)
 
     return dag
 
 
+def _fail_if_columns_overriding_functions_are_not_in_dag(
+    dag, columns_overriding_functions, is_minimal_specification
+):
+    """Fail if ``columns_overriding_functions`` are not in DAG.
+
+    TODO: Apparently, the check is pretty harsh to our tests which compute and test
```

Comment:
Since `check_minimal_specification` (forgot to rename here) is `"ignore"` by default, we can delete this todo note. The question is whether we want to require `check_minimal_specification="raise"` for our test suite.

I think it is a little bit too much. We would need to separate tests for each target because `columns_overriding_functions` and columns in `data` are target-specific and might become redundant when the target changes which will raise an error.

Reply:
Cool. Just delete.